src/main/java/core/basesyntax/service/operation/PurchaseOperation.java

```
@@ -0,0 +1,25 @@
+package core.basesyntax.service.operation;
+
+import core.basesyntax.exception.FruitShopException;
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.OperationHandler;
+import core.basesyntax.storage.Storage;
+
+public class PurchaseOperation implements OperationHandler {
+    private final Storage storage;
+
+    public PurchaseOperation(Storage storage) {
+        this.storage = storage;
+    }
+
+    @Override
+    public void handle(FruitTransaction transaction) {
+        int available = storage.get(transaction.getFruit());
+        int qty = transaction.getQuantity();
```

handle(...) lacks validation of its input. You must check `transaction != null`, `transaction.getFruit()` is non-null/non-blank, and `transaction.getQuantity() >= 0` before using them. Add precondition checks and throw FruitShopException with clear messages to fail fast on invalid input. This prevents NPEs and matches the checklist's requirement to validate inputs. Relevant method starts at line 16.